### PR TITLE
[Content] Add `astro sync` type gen command

### PR DESCRIPTION
**Pull request by bholmesdev**
## Changes

Add `astro sync` command for generating types. Includes content collections today, but could include other type gen processes in the future. Inspired by [SvelteKit sync](https://kit.svelte.dev/docs/cli#svelte-kit-sync).

## Testing

- Test that types file is written on sync with `astro:content` module declaration

## Docs

https://github.com/withastro/docs/pull/2244#pullrequestreview-1230793118

### CODE REVIEW

**Review by natemoo-re — APPROVED** [review]
Nice addition!

File: packages/astro/src/cli/sync/index.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+		contentConfigObserver: contentObservable({ status: 'loading' }),
+		logging,
+		fs,
+		settings,
+	});
+
+	try {
+		await contentTypesGenerator.init();
+	} catch (e) {
+		error(logging, 'content', 'Failed to generate content collection types: ' + e);
```
Comment (delucis) [inline packages/astro/src/cli/sync/index.ts:25]: Should new errors all now receive associated error documentation? Or is that only possible for `core`? (https://github.com/withastro/astro/blob/main/packages/astro/src/core/errors/README.md)

**Review by delucis — APPROVED** [review]
Looks ama`sync` to me! One question about errors.

File: packages/astro/src/cli/sync/index.ts (hunk 1 above)
Comment (bholmesdev) [reply packages/astro/src/cli/sync/index.ts:25]: To be honest, I'm note sure we document CLI errors as we would runtime errors! Let me raise in discord if that's alright.

Comment (bholmesdev) [reply packages/astro/src/cli/sync/index.ts]: Update: Added `8xxx` codes for CLI errors, and added a `GenerateContentTypesError` ❤️

**Comment by sarah11918** [comment]
Approving the content of the error message! (Just checking, we don't yet have any `astro sync` API docs themselves, though, right? Still TODO? That's why I'm not smashing the big Approve button.)

**Review by sarah11918 — APPROVED** [review]

**Pull request merged**